File: backend-elixir/test/unit/jwks_verifier_test.exs
defmodule BackendElixir.Auth.JwksVerifierTest do
  use ExUnit.Case, async: false

  alias BackendElixir.Auth.{JwksCacheServer, JwksVerifier}
  alias BackendElixir.TestSupport.{MockJwksServer, TestTokenHelper}

  @issuer "http://localhost:8082/realms/training"
  @audience "backend"

  setup do
    {:ok, server_info} = MockJwksServer.start()
    MockJwksServer.reset_keys()

    {:ok, cache_pid} = JwksCacheServer.start_link(jwks_url: server_info.jwks_url)

    on_exit(fn ->
      MockJwksServer.stop(server_info)
    end)

    %{server_info: server_info, cache_pid: cache_pid}
  end

  defp add_test_key(kid) do
    private_key = TestTokenHelper.generate_rsa_key_pair()
    MockJwksServer.add_key(kid, TestTokenHelper.public_jwk_fields(private_key))
    private_key
  end

  test "accepts valid token signed with known key", %{cache_pid: cache_pid} do
    private_key = add_test_key("kid-1")
    token = TestTokenHelper.make_rsa_token(private_key, "kid-1", @issuer, @audience, "99", 3600)
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)

    assert {:ok, claims} = JwksVerifier.verify(verifier, token)
    assert claims.sub == "99"
  end

  test "unknown_kid triggers refresh then succeeds if now present", %{cache_pid: cache_pid} do
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)
    # まだJWKSに何も登録されていない状態でリクエストが来た後にキーを追加する、という
    # 「未知kid -> 再取得 -> 見つかる」の流れを再現する
    private_key = TestTokenHelper.generate_rsa_key_pair()
    token = TestTokenHelper.make_rsa_token(private_key, "kid-late", @issuer, @audience, "5", 3600)
    MockJwksServer.add_key("kid-late", TestTokenHelper.public_jwk_fields(private_key))

    assert {:ok, claims} = JwksVerifier.verify(verifier, token)
    assert claims.sub == "5"
  end

  test "unknown kid still unknown after refresh is rejected", %{cache_pid: cache_pid} do
    private_key = TestTokenHelper.generate_rsa_key_pair()
    token = TestTokenHelper.make_rsa_token(private_key, "kid-never-registered", @issuer, @audience, "1", 3600)
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)

    assert {:error, :unknown_kid} = JwksVerifier.verify(verifier, token)
  end

  test "wrong issuer is rejected even with valid signature", %{cache_pid: cache_pid} do
    private_key = add_test_key("kid-2")
    token = TestTokenHelper.make_rsa_token(private_key, "kid-2", "unexpected-issuer", @audience, "1", 3600)
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)

    assert {:error, _reason} = JwksVerifier.verify(verifier, token)
  end

  test "wrong audience is rejected", %{cache_pid: cache_pid} do
    private_key = add_test_key("kid-3")
    token = TestTokenHelper.make_rsa_token(private_key, "kid-3", @issuer, "someone-else", "1", 3600)
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)

    assert {:error, _reason} = JwksVerifier.verify(verifier, token)
  end

  test "expired token is rejected", %{cache_pid: cache_pid} do
    private_key = add_test_key("kid-4")
    token = TestTokenHelper.make_rsa_token(private_key, "kid-4", @issuer, @audience, "1", -3600)
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)

    assert {:error, _reason} = JwksVerifier.verify(verifier, token)
  end

  test "rejects unexpected algorithm", %{cache_pid: cache_pid} do
    _private_key = add_test_key("kid-5")
    # RS256を期待しているVerifierに対し、HS256の(全く別の鍵体系の)トークンを送る
    token = TestTokenHelper.make_hmac_token("irrelevant-secret", @issuer, @audience, "1", 3600)
    verifier = JwksVerifier.new(cache_pid, @issuer, @audience)

    assert {:error, _reason} = JwksVerifier.verify(verifier, token)
  end
end

File: backend-js-ts-express/tests/jwks_verifier.test.ts
// backend-js-express/tests/jwks_verifier.test.jsの型付き移植。テスト内容は変更していない。
// backend-java/JwksVerifierTest.java・backend-python/test_jwks_verifier.pyの移植でもある。
// 自プロセス内蔵のモックJWKSサーバーを使い、実際のRS256署名検証・kidキャッシュ・
// 未知kid時の再取得ロジックを検証する。DB・Keycloak・bff実プロセスは不要

import test from 'node:test';
import assert from 'node:assert/strict';
import { JwksVerifier, VerifyError } from '../src/auth/jwt';
import { createMockJwksServer } from './support/mock_jwks_server';
import { generateRsaKeyPair, makeRsaToken } from './support/test_token_helper';

const ISSUER = 'https://issuer.example';

test('accepts a valid token signed with a known key', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');

    const token = makeRsaToken(privateKey, 'kid-1', ISSUER, 'backend', '99', 3600);
    const claims = await verifier.verify(token);
    assert.equal(claims.sub, '99');
    assert.equal(claims.iss, ISSUER);
  } finally {
    await mockJwks.close();
  }
});

test('an unknown kid triggers a refresh and succeeds once present', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    // JwksVerifier生成時点ではまだ鍵ゼロ件 -> 最初の検証はkid不一致で1回だけ再取得を試みる
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');
    mockJwks.addKey('kid-2', publicKey);

    const token = makeRsaToken(privateKey, 'kid-2', ISSUER, 'backend', '1', 3600);
    const claims = await verifier.verify(token);
    assert.equal(claims.sub, '1');
  } finally {
    await mockJwks.close();
  }
});

test('a kid that is still unknown after refresh is rejected', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-registered', publicKey);
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');

    // JWKSには存在しないkidで署名したトークン -> 再取得しても見つからず拒否される
    const token = makeRsaToken(privateKey, 'kid-does-not-exist', ISSUER, 'backend', '1', 3600);
    await assert.rejects(() => verifier.verify(token), VerifyError);
  } finally {
    await mockJwks.close();
  }
});

test('the wrong issuer is rejected even with a valid signature', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');

    const token = makeRsaToken(privateKey, 'kid-1', 'https://different-issuer.example', 'backend', '1', 3600);
    await assert.rejects(() => verifier.verify(token), VerifyError);
  } finally {
    await mockJwks.close();
  }
});

test('the wrong audience is rejected', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');

    const token = makeRsaToken(privateKey, 'kid-1', ISSUER, 'someone-else', '1', 3600);
    await assert.rejects(() => verifier.verify(token), VerifyError);
  } finally {
    await mockJwks.close();
  }
});

test('an expired token is rejected', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');

    const token = makeRsaToken(privateKey, 'kid-1', ISSUER, 'backend', '1', -3600);
    await assert.rejects(() => verifier.verify(token), VerifyError);
  } finally {
    await mockJwks.close();
  }
});

// アルゴリズム混同攻撃対策: ヘッダのalgがRS256以外なら拒否する
test('rejects an unexpected algorithm', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const verifier = new JwksVerifier(mockJwks.jwksUrl(), ISSUER, 'backend');

    const token = makeRsaToken(privateKey, 'kid-1', ISSUER, 'backend', '1', 3600, 'RS512');
    await assert.rejects(() => verifier.verify(token), VerifyError);
  } finally {
    await mockJwks.close();
  }
});
